regcore/db/django_models.py

```diff
@@ -1,45 +1,98 @@
 """Each of the data structures relevant to the API (regulations, notices,
 etc.), implemented using Django models"""
+import collections
+
 from django.core.exceptions import ObjectDoesNotExist
 
 from regcore.db import interface
 from regcore.models import Diff, Layer, Notice, Regulation
 
 
+def treeify(node, tree_id, pos=1, level=0):
+    """Set tree properties in memory.
+    """
+    node['tree_id'] = tree_id
+    node['level'] = level
+    node['left'] = pos
+    for child in node.get('children', []):
+        pos = treeify(child, tree_id, pos=pos + 1, level=level + 1)
+    pos = pos + 1
+    node['right'] = pos
+    return pos
+
+
+def get_adjacency_map(regs):
+    """Build mapping from node IDs to child records
+    :param regs: List of `Regulation` records
+    """
+    ret = collections.defaultdict(list)
+    for reg in regs:
+        if reg.parent_id is not None:
+            ret[reg.parent_id].append(reg)
+    return ret
+
+
+def get_id(reg, version):
+    return '{}-{}'.format(version, '-'.join(reg['label']))
+
+
 class DMRegulations(interface.Regulations):
     """Implementation of Django-models as regulations backend"""
     def get(self, label, version):
         """Find the regulation label + version"""
-        try:
-            reg = Regulation.objects.get(version=version,
-                                         label_string=label)
-            as_dict = {
-                'label': reg.label_string.split('-'),
-                'text': reg.text,
-                'node_type': reg.node_type,
-                'children': reg.children
-            }
-            if reg.title:
-                as_dict['title'] = reg.title
-            return as_dict
-        except ObjectDoesNotExist:
+        regs = Regulation.objects.filter(
+            version=version,
+            label_string=label,
+        ).get_descendants(
+            include_self=True,
+        )
+        regs = list(regs.all())
+        if not regs:
             return None
+        adjacency_map = get_adjacency_map(regs)
+        return self._serialize(regs[0], adjacency_map)
```

Comment:
I don't think it would overwrite the `children` manager (which is completely unused in our use case), but I _would_ expect the library to have a mechanism to return a fully-nested representation.

:100: to PRing upstream. I just need to sit down and implement :p